### PR TITLE
Rotation aug  implementation

### PR DESCRIPTION
**Pull request by tianyu-z**
I guess the most weird part of the rotation implementation is to rotate the segmentation pts. This is the logic of how it works. Hope this graph can help you review the code.
image

### CODE REVIEW

**Comment by tianyu-z** [comment]
Besides, an option for HRNet Normalization is added. Please check opts.data.normalization in defaults.yaml

File: omnigan/losses.py [hunk 1]
```diff
@@ -466,7 +479,7 @@ def __call__(self, prediction, target):
 
 class ADVENTAdversarialLoss(nn.Module):
     """
-        TODO
+        The first and second arguement are tensor. The third arguement is a discriminator model.
```
Comment (melisandeteng) [inline omnigan/losses.py:482]: typo "argument"

File: omnigan/losses.py [hunk 2]
```diff
@@ -276,14 +276,16 @@ def __call__(self, prediction, target):
         relDiff = torch.mean(d) * torch.mean(d)
         return diff - relDiff
 
+
 class SIGMLoss(nn.Module):
-    def __init__(self, gmweight, device='cuda'):
+    def __init__(self, gmweight, device="cuda"):
```
Comment (melisandeteng) [inline omnigan/losses.py:281]: can you merge master into your branch before we merge yours ? this loss is outdated

File: omnigan/transforms.py [hunk 3]
```diff
@@ -81,6 +83,83 @@ def __call__(self, data):
...
+            tanAngle = np.tan(angle_selected)
+        if len(rotated.shape) == 3:
+            rotated = rotated.unsqueeze(0)
+        _, _, h, w = rotated.shape
+        a = int(h * tanAngle / (tanAngle + 1))
+        c = int(w * tanAngle / (tanAngle + 1))
+        return rotated[:, :, c:-c, a:-a]
+
+    def mapping(self, rotated):
+        Dict = {
```
Comment (melisandeteng) [inline omnigan/transforms.py:107]: where do the values come from ?

File: omnigan/transforms.py [hunk 4]
```diff
@@ -81,6 +83,83 @@ def __call__(self, data):
         return {task: torch.flip(tensor, [3]) for task, tensor in data.items()}
 
 
+class RandomRotations:
+    def __init__(self, p=0.5, angle=5):
+        self.p = p
+        self.angle = angle
+
+    def cut_balck_edge(self, rotated, angle_selected):
```
Comment (melisandeteng) [inline omnigan/transforms.py:91]: so that means you crop so that we don't have the black corners, whatever the rotation is ?

**Review by melisandeteng — APPROVED** [review]

File: omnigan/transforms.py (hunk 4 above)
Comment (tianyu-z) [reply omnigan/transforms.py]: Indeed~

File: omnigan/transforms.py (hunk 3 above)
Comment (tianyu-z) [reply omnigan/transforms.py]: By manual matching Lol

File: omnigan/losses.py (hunk 2 above)
Comment (tianyu-z) [reply omnigan/losses.py]: Done~

File: omnigan/losses.py (hunk 1 above)
Comment (tianyu-z) [reply omnigan/losses.py]: Done. Thx!

**Comment by tianyu-z** [comment]
Got some issues after merging master into this branch. Trying to fix......

File: omnigan/transforms.py (hunk 3 above)
Comment (melisandeteng) [reply omnigan/transforms.py]: but what do they correspond to ? Im just curious :)

File: omnigan/blocks.py [hunk 5]
```diff
@@ -297,35 +297,40 @@ class DepthDecoder(nn.Module):
     """#Depth decoder based on depth auxiliary task in DADA paper
 
     """
+
```
Comment (vict0rsch) [inline omnigan/blocks.py:300]: Do these changes mean you did not blackify @melisandeteng ?

File: omnigan/data.py [hunk 6]
```diff
@@ -380,7 +380,7 @@ def get_loader(mode, domain, opts):
             mode, domain, opts, transform=transforms.Compose(get_transforms(opts))
         ),
         batch_size=opts.data.loaders.get("batch_size", 4),
-        shuffle=True,
+        shuffle=opts.data.loaders.shuffle,
```
Comment (vict0rsch) [inline omnigan/data.py:383]: I think we should always shuffle that's why I had removed that option. Validation images logged in comet are not created from iterating over the val loaders, just stored once before training into the `display_images`

File: omnigan/transforms.py [hunk 7]
```diff
@@ -89,7 +170,7 @@ def __init__(self):
     def __call__(self, data):
         new_data = {}
         for task, im in data.items():
-            if task in {"x", "a"}:
+            if task in {"x"}:
```
Comment (vict0rsch) [inline omnigan/transforms.py:173]: rephrase `task == "x"`

File: omnigan/transforms.py [hunk 8]
```diff
@@ -140,6 +225,9 @@ def get_transform(transform_item):
     if transform_item.name == "hflip" and not transform_item.ignore:
         return RandomHorizontalFlip(p=transform_item.p or 0.5)
 
+    if transform_item.name == "rot" and not transform_item.ignore:
```
Comment (vict0rsch) [inline omnigan/transforms.py:228]: I think `rotation` is a good enough name :)

File: omnigan/transforms.py [hunk 9]
```diff
@@ -81,6 +83,85 @@ def __call__(self, data):
         return {task: torch.flip(tensor, [3]) for task, tensor in data.items()}
 
 
+class RandomRotations:
+    def __init__(self, p=0.5, angle=5):
```
Comment (vict0rsch) [inline omnigan/transforms.py:87]: This needs way more comments and docstrings explaining your logic for  the PR to be mergeable.

Especially (but not only): `seg_dict`, `mapping(...)`, difference between `s` and other tasks, why go to tensor then to pil then to numpy then back to tensor and what not

**Review by vict0rsch — CHANGES_REQUESTED** [review]

**Comment by vict0rsch** [comment]
@tianyu-z I understood from your presentation that rotation tools induce noise in the pixel values.

Have you researched extensively existing tools and techniques? Other packages like `skimage` or `opencv`? Tried other types for the data _before_ the rotation?

If so, then we should keep your current implementation as it seems to work.

I'd be interested in measuring the loading time overhead: can you check the time this particular transform takes on an image (vs other transforms, to understand if this adds 20% calculations or 2000%)?

**Comment by tianyu-z** [comment]
> @tianyu-z I understood from your presentation that rotation tools induce noise in the pixel values.
> 
> Have you researched extensively existing tools and techniques? Other packages like `skimage` or `opencv`? Tried other types for the data _before�_ the rotation?
> 
> If so, then we should keep your current implementation as it seems to work.
> 
> I'd be interested in measuring the loading time overhead: can you check the time this particular transform takes on an image (vs other transforms, to understand if this adds 20% calculations or 2000%)?

@vict0rsch I tried skimage.transform.rotate which introduced even much noise than the scipy one which is not acceptable in segmentation data. 
In detail: 
In the top20 Counter.keys, we can see some kind of numbers like 1.9999996 in Scipy and numbers like 7.4xxx are very rare in Scipy rotator. For skimage.transform.rotator, it gives a 6.45xxx in top6 Counter.keys which I think is pretty bad.
 
For cv2, I think they only provide rotations of multiplication of 90 degrees. Torch rotation is really slow when it works on normal pictures, but it's good on segmentation data.

When it comes to time elapsed, I was very careful about it. Hope everything will be good. I am running an exp with norot and an exp with rot. For bs = 6, in the rot exp I got step time ranging from 4.87 to 5.23 (The prob to trigger rot was set to 0.8.). In the norot exp, I got step time raning from 4.80 to 5.12. I will double-check it after I get up.

**Comment by tianyu-z** [comment]
An example about the results.
image

**Comment by tianyu-z** [comment]
rot exp: https://www.comet.ml/tianyu-z/omnigan/beab426b94c04843a86696f6a2e15eac?experiment-tab=metrics
no rot exp: https://www.comet.ml/tianyu-z/omnigan/f6e53d7ea74f47deb488dab58f6e2431?experiment-tab=metrics

**Comment by vict0rsch** [comment]
Ok perfect, GPU utilization is still close to 100% so I'm good with this, you just need to add comments and docstrings  before we can merge

**Comment by tianyu-z** [comment]
> Ok perfect, GPU utilization is still close to 100% so I'm good with this, you just need to add comments and docstrings  before we can merge

Thanks!
